### PR TITLE
Stylo: Add support for grid-template-{rows,columns}

### LINKED ISSUE #15311: Implement parsing/serialization for `grid-template-{row,columns}`
They [both](https://drafts.csswg.org/css-grid/#propdef-grid-template-rows) take `none | <track-list> | <auto-track-list>` (the parsing code for "track list" thingy may be a nightmare).

### PR DESCRIPTION
**Pull request by ghost**
This has the implementation of grid's `<track-list>` nightmare.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15311

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16067)

### CODE REVIEW

**Comment by Wafflespeanut** [comment]
cc @Manishearth @emilio @canaltinova @upsuper

**Comment by Wafflespeanut** [comment]
This is now ready for review!

**Comment by Wafflespeanut** [comment]
Now, I'm specifically concerned about [this](https://github.com/Wafflespeanut/servo/blob/26b82834504a4bb780ec0f95a2e7c98e8bbd73b1/components/style/values/specified/grid.rs#L438) and [this](https://github.com/Wafflespeanut/servo/blob/26b82834504a4bb780ec0f95a2e7c98e8bbd73b1/components/style/values/specified/grid.rs#L589), which make the struct rather big.

File: components/style/values/specified/grid.rs [hunk 1]
```diff
@@ -408,3 +408,150 @@ impl ToCss for TrackRepeatFirst {
...
+}
+
+/// The structure corresponding to the various `repeat` functions.
+///
+/// https://drafts.csswg.org/css-grid/#typedef-track-repeat
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct TrackRepeat<L> {
+    /// The type of `repeat` (based on the values we encountered in parsing)
+    pub repeat_type: TrackRepeatType,
```
Comment (Manishearth) [inline components/style/values/specified/grid.rs]: do we actually need to store this field? AIUI it can be derived from the values of the other fields.

File: components/style/values/specified/grid.rs [hunk 2]
```diff
@@ -555,3 +555,169 @@ impl<L: ToComputedValue> ToComputedValue for TrackRepeat<L> {
...
+///
+/// https://drafts.csswg.org/css-grid/#typedef-track-list
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct TrackList<L> {
+    /// The type of this `<track-list>` (auto, explicit or general).
+    ///
+    /// In order to avoid parsing the same value multiple times, this does a single traversal
+    /// and arrives at the type of value it has parsed (or bails out gracefully with an error).
+    pub list_type: TrackListType,
```
Comment (Manishearth) [inline components/style/values/specified/grid.rs]: again, seems to be something that can be derived.

**Review by Manishearth — CHANGES_REQUESTED** [review]
mostly looking good, will have to have a closer look at list and repeat parsing.

**Comment by Manishearth** [comment]
geckotry imported patch at https://treeherder.mozilla.org/#/jobs?repo=try&revision=bae57da326dac72a88b98da6cf87def4e9f614ac

File: components/style/values/specified/grid.rs (hunk 1 above)
Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: We could say that it's an `<auto-repeat>` if we have `auto-fill | auto-fit` in the first argument. But, when it comes to `<fixed-repeat>`, we may have to check all the items in the list to make sure that they're all `<fixed-size>` (and this will happen every time we parse a `repeat()` in `<track-list>`).

File: components/style/values/specified/grid.rs (hunk 2 above)
Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: Yes, we can determine whether it's a `<track-list>` or `<auto-track-list>` from the optional index. Right now, we don't have much use for `<explicit-track-list>`, it's used only in `grid-template` shorthand. Not sure whether it's okay to check all those list items later.

**Comment by Manishearth** [comment]
oh, my bad, you haven't done the gecko glue yet :P

**Comment by Wafflespeanut** [comment]
Now, I've moved the optional index into `TrackListType`, so we won't be needing the additional field,

File: components/style/values/specified/grid.rs (hunk 1 above)
Comment (Manishearth) [inline components/style/values/specified/grid.rs]: But --- that's all something that matters when parsing, once parsed you don't need to store the fact that it's a particular type, no?

File: components/style/values/specified/grid.rs (hunk 2 above)
Comment (Manishearth) [inline components/style/values/specified/grid.rs]: The thing is we never need to determine this, so why store it? That's just the way the grammar is structured.

Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: Note that the `Option<u32>` (which marks the index of `<auto-repeat>`) is now stored inside `TrackListType`, both have the same sizes, so this shouldn't matter now?

File: components/style/values/specified/grid.rs (hunk 1 above)
Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: I understand that this is only required for parsing. But, we need to know about the type of `TrackRepeat` *while* parsing `TrackList`. Maybe we should drop the `Parse` impl and have a built-in method for returning `TrackRepeat` and its type.

Comment (Manishearth) [inline components/style/values/specified/grid.rs]: right, so have it be a local variable during parsing, no?

Comment (Manishearth) [inline components/style/values/specified/grid.rs]: oh, I see how you're using it. I think that function is better served by having two parse functions. And it definitely doesn't need to be in the computed value.

File: components/style/values/specified/grid.rs [hunk 3]
```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
...
+}
+
+/// The initial argument of the `repeat` function.
+///
+/// https://drafts.csswg.org/css-grid/#typedef-track-repeat
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum RepeatCount {
+    /// A positive integer. This is allowed only for `<track-repeat>` and `<fixed-repeat>`
+    Number(u32),
```
Comment (stshine) [inline components/style/values/specified/grid.rs]: I suppose there is no need to convert to `u32` here?

Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: Oh, that's just to be consistent with a CSS integer (since it's an index, it doesn't have to be signed, and `usize` is platform-specific, so something definitive).

File: components/style/values/specified/grid.rs [hunk 4]
```diff
@@ -46,15 +46,15 @@ impl ToCss for GridLine {
...
         }
 
         if let Some(i) = self.integer {
-            try!(write!(dest, " {}", i));
+            write!(dest, " {}", i)?;
         }
 
         if let Some(ref s) = self.ident {
-            try!(write!(dest, " {}", s));
+            write!(dest, " {}", s)?;
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: This needs to be escaped as a custom ident.

File: components/style/values/specified/grid.rs [hunk 5]
```diff
@@ -46,15 +46,15 @@ impl ToCss for GridLine {
         }
 
         if self.is_span {
-            try!(dest.write_str("span"));
+            dest.write_str("span")?;
         }
 
         if let Some(i) = self.integer {
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: This one needs to be `Integer` (assuming it's the specified value), in order to correctly handle `calc()`.

File: components/style/values/specified/grid.rs [hunk 6]
```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
...
+
+            size.to_css(dest)?;
+            if i < total - 1 {
+                dest.write_str(" ")?;
+            }
+        }
+
+        if !self.line_names_last.is_empty() {
+            dest.write_str(" [")?;
+            dest.write_str(&self.line_names_last.join(" "))?;
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: let's just iterate over line_names_last in order to avoid allocating? Seems easy enough.

File: components/style/values/specified/grid.rs [hunk 7]
```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
...
+impl<L: ToCss> ToCss for TrackList<L> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        for (i, &(ref names, ref value)) in self.values_list.iter().enumerate() {
+            if i > 0 {
+                dest.write_str(" ")?;
+            }
+
+            if !names.is_empty() {
+                dest.write_str("[")?;
+                dest.write_str(&names.join(" "))?;
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: ditto.

File: components/style/values/specified/grid.rs [hunk 8]
```diff
@@ -301,3 +340,377 @@ impl<L: ToComputedValue> ToComputedValue for TrackSize<L> {
...
+                dest.write_str("] ")?;
+            }
+
+            value.to_css(dest)?;
+        }
+
+        if !self.line_names_last.is_empty() {
+            dest.write_str(" [")?;
+            dest.write_str(&self.line_names_last.join(" "))?;
```
Comment (emilio) [inline components/style/values/specified/grid.rs]: ditto. Also, they need to be escaped I believe. Please add a helper function to do so.

**Review by emilio — CHANGES_REQUESTED** [review]
I haven't seen the rest of the grid code, but please do check for similar bugs.

**Comment by Wafflespeanut** [comment]
This is ready for review! cc @servo/style (whoo, team) 😄

**Comment by Wafflespeanut** [comment]
It's blocked on [Bug 1354775](https://bugzilla.mozilla.org/show_bug.cgi?id=1354775) though.

File: components/style/values/specified/grid.rs [hunk 9]
```diff
@@ -406,9 +426,147 @@ impl ToCss for RepeatCount {
...
+/// The type of `repeat` function (only used in parsing).
+///
+/// https://drafts.csswg.org/css-grid/#typedef-track-repeat
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+enum RepeatType {
+    /// [`<track-repeat>`](https://drafts.csswg.org/css-grid/#typedef-track-repeat)
+    Auto,
+    /// [`<auto-repeat>`](https://drafts.csswg.org/css-grid/#typedef-auto-repeat)
+    Normal,
```
Comment (canaltinova) [inline components/style/values/specified/grid.rs]: Links of these two variants looks opposite. Aren't they?

Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: Good catch! :)

**Comment by Wafflespeanut** [comment]
A bunch of tests seem to pass - https://treeherder.mozilla.org/#/jobs?repo=try&revision=c4d29aae1d2b293acfae9d3120072a7e5cb74ec5&selectedJob=97141102 (but there are crashes in some reftests too)

**Comment by Wafflespeanut** [comment]
The [recent try build](https://treeherder.mozilla.org/#/jobs?repo=try&revision=abba7c05ca8a8a367bfeb706c9c28b2b72230029) shows a lot of passing mochitests and reftests. One crashtest is still failing (on some assertion error, which I thought I'd fixed). I'll look into it tomorrow. Meanwhile, it'd be nice if I get some eyes on this :)

**Comment by Wafflespeanut** [comment]
Try build succeeds - https://treeherder.mozilla.org/#/jobs?repo=try&revision=0dc58d10bbeb848412baabc952c91289fca70575

**Comment by Wafflespeanut** [comment]
Wrapping the `nsTArray<nsString>` inside a struct had some trouble with the hazard analysis suite. Now, the thing is just a pointer, and so, `nsTArray` needs to be imported (for which the build script needs to be changed). So, this needs coordinated landing.

Here's the try build for hazard stuff and stylo debug (of both the patches combined) - https://treeherder.mozilla.org/#/jobs?repo=try&revision=8ca319166ea7b5f9c2ad69b347007ef787d80f7d

**Comment by nox** [comment]
I feel like this patch piggybacks multiple different properties with different kinds of values in a single Rust enum. Am I misunderstanding things?

**Review by nox — COMMENTED** [review]
Two remarks.

File: components/style/values/specified/grid.rs [hunk 10]
```diff
@@ -367,3 +367,48 @@ pub fn parse_line_names(input: &mut Parser) -> Result<Vec<String>, ()> {
...
 }
+
+/// The initial argument of the `repeat` function.
+///
+/// https://drafts.csswg.org/css-grid/#typedef-track-repeat
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum RepeatCount {
+    /// A positive integer. This is allowed only for `<track-repeat>` and `<fixed-repeat>`
+    Number(u16),
```
Comment (Wafflespeanut) [inline components/style/values/specified/grid.rs]: You're right. The funny thing is that gecko doesn't support it - https://bugzilla.mozilla.org/show_bug.cgi?format=default&id=1350069 😄

Comment (nox) [inline components/style/values/specified/grid.rs]: Let's support it anyway, and worst case we will dumb it down when time has come.

**Comment by Wafflespeanut** [comment]
Addressed review comments! (I've been waiting for this for a month now!)

Finally!!! @bors-servo r=nox

**Comment by nox** [comment]
@bors-servo p=8

Sorry.
